src/imperative/imperative_utils.cc

```diff
@@ -22,6 +22,106 @@
 
 namespace mxnet {
 namespace imperative {
+
+inline std::vector<NDArray*> NodeInputs(const nnvm::IndexedGraph& idx,
+                                        const int node_idx,
+                                        const std::vector<NDArray*> arrays) {
+  const nnvm::IndexedGraph::Node& node = idx[node_idx];
+  const size_t num_inputs = node.inputs.size();
+  std::vector<NDArray*> ndinputs;
+  ndinputs.reserve(num_inputs);
+  for (const auto& j : node.inputs) {
+    size_t eid = idx.entry_id(j);
+    ndinputs.emplace_back(arrays[eid]);
+  }
+  return ndinputs;
+}
+
+inline std::vector<NDArray*> NodeOutputs(const nnvm::IndexedGraph& idx,
+                                         const int node_idx,
+                                         const std::vector<NDArray*> arrays) {
+  const nnvm::IndexedGraph::Node& node = idx[node_idx];
+  const size_t num_outputs = node.source->num_outputs();
+  std::vector<NDArray*> ndoutputs;
+  ndoutputs.reserve(num_outputs);
+  for (size_t j = 0; j < num_outputs; ++j) {
+    size_t eid = idx.entry_id(node_idx, j);
+    ndoutputs.emplace_back(arrays[eid]);
+  }
+  return ndoutputs;
+}
+
+inline std::vector<OpReqType> NodeReq(const nnvm::IndexedGraph& idx,
+                                      const int node_idx,
+                                      const std::vector<OpReqType> array_reqs) {
+  const nnvm::IndexedGraph::Node& node = idx[node_idx];
+  const size_t num_outputs = node.source->num_outputs();
+  std::vector<OpReqType> req;
+  req.reserve(num_outputs);
+  for (size_t j = 0; j < num_outputs; ++j) {
+    size_t eid = idx.entry_id(node_idx, j);
+    req.push_back(array_reqs[eid]);
+  }
+  return req;
+}
+
+inline void InvokeOperator(const nnvm::IndexedGraph& idx,
```

Comment:
Why inline?

Reply:
Also this function should be static or in the anonymous namespace as it's not used outside this unit.